plutus-core/plutus-core/src/PlutusCore/Parser/ParserCommon.hs

```diff
@@ -99,6 +119,24 @@ wordPos ::
     T.Text -> Parser SourcePos
 wordPos w = lexeme $ try $ getSourcePos <* symbol w
 
+-- | Given a `Parser` that captures the ending position, supply the starting position
+-- to make it a `SrcSpan`.
+withSpan :: (SrcSpan -> a -> b) -> Parser (a, SourcePos) -> Parser b
```

Comment:
So, this seemed a bit clumsy to me. I thought about it a bit, and indeed, `ParsecT` has an instance of `MonadFix`. So we can write this with `RecursiveDo`:
```
withSpan :: (SrcSpan -> Parser a) -> Parser a
withSpan' f = mdo
  start <- getSourcePos
  res <- f spn
  end <- getSourcePos
  let spn = toSrcSpan start end
  pure res

lamTerm :: Parser PTerm
lamTerm =
    withSpan
        (\sp -> inParens (UPLC.LamAbs sp <$> (symbol "lam" *> name) <*> term))
```
Note that it's using the normal `inParens` etc. I think this is very nice and clean and expresses exactly what it is that we want to do, while being minimally invasive to everything else (all `Parser`s always return things with `SrcSpan`s in them).

And yes, this works, I tried it. So long as the argument to `withSpan` is lazy, which it always is.

Reply:
But `type Parser = ParsecT ParserError T.Text (StateT ParserState Quote)`, and `Quote` doesn't have a `MonadFix` instance

Also the normal `inParens` consumes trailing whitespaces, so I assume you meant the version of `inParens` that doesn't.

Reply:
Heh, Ok, I didn't know `StateT` has `MonadFix` instances. I've never used `MonadFix` in my life :stuck_out_tongue:

Reply:
This is also the first time I've used it! But it does seem like a perfect usecase!

Reply:
Updated. This is indeed a great use case of `mdo`!